bootstrap/pkg/kfapp/coordinator/coordinator.go

```diff
@@ -774,24 +783,32 @@ func (kfapp *coordinator) Apply(resources kftypesv3.ResourceEnum) error {
 		}
 	}
 
+	// TODO(gabrielwen): Move `gcpAddedConfig` back to gcp.go.
 	switch resources {
 	case kftypesv3.ALL:
-		if err := platform(); err != nil {
-			return err
-		}
-		if err := k8s(); err != nil {
-			return err
-		}
-		return gcpAddedConfig()
+		return simpleReconcile(func() error {
```

Comment:
Should this loop be implemented in Apply?
https://github.com/kubeflow/kubeflow/blob/7789ce18a468b255739b62ada53ecf14924b37a0/bootstrap/pkg/kfapp/coordinator/coordinator.go#L777

Also I think you need to think through the exit conditions and continuation behavior.

e.g. how does Platform () indicate that it has reached a terminal state and doesn't need to be invoked again?

Consider the following use case. The GCP plugin wants to have logic that can only be executed once the K8s namespace exists.

So the first pass through the loop the GCP plugin/app would deploy a GKE cluster. But since the namespace and Kubeflow haven't been deployed yet; we can't initialize a default profile in a GCP specific way.

If it returns an error then simpleRecocile would return and we would just retry calling platform(). But that isn't what we want because the namespace still won't exist.

So we need to call k8s() which should succeed and then we want to retry.

Reply:
sounds good, let me modify this.

Reply:
@jlewi could you take another look?  I tried to make it simple unless you want me to implement the full version of Reconcile in v1beta1.